### PR TITLE
introducer Signer and Verifier interfaces

### PR DESCRIPTION
**Pull request by imjasonh**
This should help us plug in other signing/verifying implementations, possibly backed by KMS systems, in the future.

Callers of `SignIndex` will need to call `NewKeySignerVerifier` and pass that to `SignIndex`, instead of just passing the filename of the private signing key.

### CODE REVIEW

**Comment by imjasonh** [comment]
I wonder whether this would also be a convenient time to introduce a `context.Context` to `Sign` and `Verify`, if we think there might be implementations that would take advantage of it. It's not going to get easier to add it than now.

File: pkg/apk/sign.go [hunk 1]
```diff
@@ -32,8 +37,119 @@ import (
...
+	passphrase      string
+}
+
+func (s *keySignerVerifier) KeyName() string {
+	return filepath.Base(s.privkeyFile)
+}
+
+func (s *keySignerVerifier) Sign(data []byte) ([]byte, error) {
+	if len(data) != sha1.Size {
+		return nil, errDigestNotSHA1
```
Comment (mattmoor) [inline pkg/apk/sign.go:96]: Some KMS interfaces support taking the data and hashing itself, and some take the data pre-hashed.  Can we adapt the method name and argument to make it clearer that this is taking/signing the SHA1?

**Review by deitch — CHANGES_REQUESTED** [review]
I really like what you did here @imjasonh . Actually, to the point where I am not even sure it is fair to ask for these changes. I will ask anyways, but if you push back and think they are unreasonable, we can approve and merge it in.

* I think the whole sign/verify thing should go in its own path. I know it wasn't before, but that was because it was constrained into apko's structure. Now, it has the space to do what it needs. Maybe `pkg/apk/signatures` or something like that?
* Once we do, can we put the interfaces (`Signer`, `Verifier`, `SignerVerifier`) into their own files, and then the actual impl into its own as well
* The specific `keySignerVerifier` we have here is specifically for our current needs, i.e. RSA signatures with a SHA1 digest. Should we reflect that somehow? The only reason I am not 100% sure we should is that is what is used with apk indexes for now (but that might change).

For the latter point, we either could have a separate implementation, e.g.:

```go
type rsaWithSha1KeySignerVerifier struct
...
NewRSAWithSha1KeySignerVerifier()
```

Or a generic keySignerVerifier that recognizes the formats it is passed. That could be either because it knows how to read the keys and hashes (which is a little tricky) or is passed options to `New()`.

**Comment by deitch** [comment]
@imjasonh I may have complicated your PR here, my apologies. I had to move some of the signing stuff around in order to be able to extract it from melange (#15 and some subsequent ones). The structure hasn't really changed, just the location and some minor parameters. So your PR still should work, it just needs a little more than a simple rebase.

**Comment by deitch** [comment]
@imjasonh any desire to get this one in?

**Pull request closed without merging**